rspace/src/test/scala/coop/rchain/rspace/history/HistoryActionTests.scala

```diff
@@ -0,0 +1,340 @@
+      val deleteTwo       = delete(insertTwo.head.key) :: delete(insertTwo(1).key) :: Nil
+      val deleteOneAndTwo = deleteOne ::: deleteTwo
+
+      for {
+        emptyHistory <- emptyHistoryF
+
+        historyOne       <- emptyHistory.process(insertOne)
+        historyTwo       <- emptyHistory.process(insertTwo)
+        historyOneAndTwo <- emptyHistory.process(insertOneAndTwo)
+
+        historyOneAndTwoAnotherWay <- historyOne.process(insertTwo)
+        _                          = historyOneAndTwo.root shouldBe historyOneAndTwoAnotherWay.root
+
+        historyOneAnotherWay <- historyOneAndTwo.process(deleteTwo)
+        _                    = historyOne.root shouldBe historyOneAnotherWay.root
+
+        historyTwoAnotherWay <- historyOneAndTwo.process(deleteOne)
+        _                    = historyTwo.root shouldBe historyTwoAnotherWay.root
+
+        emptyHistoryAnotherWay <- historyOneAndTwo.process(deleteOneAndTwo)
+        _                      = emptyHistory.root shouldBe emptyHistoryAnotherWay.root
+
+      } yield ()
+  }
+
+  "adding already existing records" should "not change history" in withEmptyHistoryAndStore {
+    (emptyHistoryF, inMemoStore) =>
+      val inserts = insert(_zeros) :: Nil
+      for {
+        emptyHistory     <- emptyHistoryF
+        emptyHistorySize = inMemoStore.sizeBytes()
+        historyOne       <- emptyHistory.process(inserts)
+        historyOneSize   = inMemoStore.sizeBytes()
+        historyTwo       <- historyOne.process(inserts)
+        historyTwoSize   = inMemoStore.sizeBytes()
+        _                = historyOne.root shouldBe historyTwo.root
+        _                = emptyHistorySize shouldBe 0L
+        _                = historyOneSize shouldBe historyTwoSize
+      } yield ()
+  }
+// TODO: Don't works for MergingHistory
+  "collision detecting in KVDB" should "works" in withEmptyHistoryAndStore {
+    (emptyHistoryF, inMemoStore) =>
+      def copyBVToBuf(bv: ByteVector): ByteBuffer = {
+        val arr    = bv.toArray
+        val newBuf = ByteBuffer.allocateDirect(arr.length)
+        newBuf.put(arr).rewind()
+      }
+      val insertRecord    = insert(_zeros) :: Nil
+      val deleteRecord    = delete(_zeros) :: Nil
+      val collisionKVPair = (copyBVToBuf(History.emptyRootHash.bytes), randomBlake.bytes)
+      for {
+        emptyHistory <- emptyHistoryF
+        newHistory   <- emptyHistory.process(insertRecord)
+        _            <- inMemoStore.put[ByteVector](Seq(collisionKVPair), copyBVToBuf)
+        err          <- newHistory.process(deleteRecord).attempt
+      } yield {
+        err.isLeft shouldBe true
+        val ex = err.left.get
+        ex shouldBe a[RuntimeException]
+        ex.getMessage shouldBe
+          s"1 collisions in KVDB (first collision with key = " +
+            s"${History.emptyRootHash.bytes.toHex})."
+      }
+  }
+
+  "randomly insert or delete" should "return the correct result" in withEmptyHistory {
```

Comment:
I added a random insert and delete test. Hope you don't mind. :)

Reply:
This one is very fast compared to the `MergeHistory` one. Very excellent and obvious performance gain!!